talkspace/chatGPT.js

```diff
@@ -0,0 +1,295 @@
+      },
+    });
+
+    // Retrieve monthly credits used by the specified patient
+    const stats = await Booking.findAll({
+      attributes: [
+        [sequelize.fn('SUM', sequelize.literal('CASE WHEN "Booking"."status" = "confirmed" THEN "Credit"."type" END')), 'totalCreditsUsed'],
+        [sequelize.fn('MONTH', sequelize.col('"Booking"."time"')), 'month'],
+        [sequelize.fn('YEAR', sequelize.col('"Booking"."time"')), 'year'],
+      ],
+      include: [
+        {
+          model: Credit,
+          attributes: [],
+          where: {
+            BookingId: sequelize.literal('"Booking"."id"'), // Match Credit to Booking
+          },
+        },
+      ],
+      where: {
+        patient: patientId,
+        status: 'confirmed',
+      },
+      group: ['month', 'year'],
+    });
+
+    // Extract the results from the stats
+    const result = stats.map((row) => ({
+      totalCreditsUsed: row.getDataValue('totalCreditsUsed') || 0,
+      month: row.getDataValue('month'),
+      year: row.getDataValue('year'),
+    }));
+
+    // Calculate the percentage for each month
+    const totalCreditsAvailable = totalCreditsQuery || 1; // To avoid division by zero
+    const monthlyStatsWithPercentage = result.map((row) => ({
+      ...row,
+      percentageCreditsUsed: (row.totalCreditsUsed / totalCreditsAvailable) * 100,
+    }));
+
+    return monthlyStatsWithPercentage;
+  } catch (error) {
+    console.error('Error getting monthly credits used statistics:', error);
+    throw error;
+  }
+}
+
+
+
+// Endpoint to create a booking with an unused credit
+app.post('/bookings', async (req: Request, res: Response) => {
+  const { time, patient, provider } = req.body;
+
+  try {
+    // Find an unused credit that is not expired
+    const d = new Date();
+    const credit = await Credit.findOne({
+      where: {
+        expirationDate: {
+          [sequelize.Op.gt]: d, // Expiration date is greater than the current date
+        },
+        BookingId: null, // Credit is not associated with any booking
+      },
+    });
+
+    if (!credit) {
+      return res.status(404).json({ error: 'No unused, non-expired credits found.' });
+    }
+
+    // Create a booking associated with the credit
+    const booking = await Booking.create({ time, patient, provider });
+
+    // Record the initial status in the booking status history
+    await BookingStatusHistory.create({ status, BookingId: booking.id });
+
+    
+    // Associate the booking with the credit
+    await booking.setCredit(credit);
+
+    res.status(201).json({
+      message: 'Booking created successfully',
+      booking: booking.toJSON(),
+    });
+  } catch (error) {
+    console.error('Error creating booking:', error);
+    res.status(500).json({ error: 'An error occurred while creating the booking.' });
+  }
+});
+
+
+// Endpoint to retrieve bookings for a specific user (patient or provider)
+app.get('/bookings', async (req: Request, res: Response) => {
```

Comment:
Please try to implement the same as the last comments fro this function content.